### ISSUE TITLE
TASK: Remove hard-coded paths etc

### ISSUE THREAD

**Issue by HenrikBengtsson** [UCSF-Costello-Lab/LG3_Pipeline#3]
As part of Ziv Lab needs to run LG3 themselves, I'm going to replace all hard-coded paths in the scripts with environment variables.  (I did this back in 2015, but I have to redo it by hand because the LG3_Pipeline production pipeline has since been updated).

I will do this migration in a separate branch and I'm hoping that we can merge my updates into the production pipeline so we can keep one single master code base.

**Comment by HenrikBengtsson** [comment]
Idea is to update all scripts to set up env vars, at the top, which backward-compatible default values:
```bash
LG3_HOME=${LG3_HOME:-/home/jocostello/shared/LG3_Pipeline}
```
then replace all occurrences such as:
```bash
BWA=/home/jocostello/shared/LG3_Pipeline/tools/bwa-0.5.10/bwa
BWA_INDEX="/home/jocostello/shared/LG3_Pipeline/resources/bwa_indices/hg19.bwa"
```
with
```bash
BWA="${LG3_HOME}/tools/bwa-0.5.10/bwa"
BWA_INDEX="${LG3_HOME}/resources/bwa_indices/hg19.bwa"
```

I understand that eventually we'll end up with wanting to have a central config file (as in LG3 v2), but let us discuss that when we get there.  I just want to make minimal, backward compatible, changes to the existing code, while allowing to host/run/input/output elsewhere.

**Comment by HenrikBengtsson** [comment]
For the sanity of this task, I think it's important to fix the ShellCheck warnings/errors first, test those changes and bring that into production.  Then one can start on replacing hard-coded strings with env vars.

**Comment by ivan108** [comment]
OK, I will work with shellcheck directly on production code for now...

________________________________

**Comment by HenrikBengtsson** [comment]
@ivan108, it looks like you've completed [testing/validation](#6) of Steps 1-6b in [Readme_demo.md](https://github.com/UCSF-Costello-Lab/LG3_Pipeline/blob/develop/Readme_demo.md).  Does this mean that those steps have been fully verified, or will you do some manual/visual validation of the final results before concluding things works as before?

If no objections, I'll start working on this task of migrating hardcoded paths into environment variables, e.g. updating
```sh
/home/jocostello/shared/LG3_Pipeline/scripts/trim_galore.sh -quality="$QTY" "$FASTQ1" "$FASTQ2" || { echo "ERROR: trim_g
alore FAILED"; exit 1; }
```
to
```sh
LG3_HOME=${LG3_HOME:-/home/jocostello/shared/LG3_Pipeline}
"${LG3_HOME}/scripts/trim_galore.sh" -quality="$QTY" "$FASTQ1" "$FASTQ2" || { echo "ERROR: trim_g
alore FAILED"; exit 1; }
```

PS. I will work toward a seperate branch, so I won't make any changes to the master or develop branches.

**Comment by ivan108** [comment]
I have saved the results from the original run, so I am/will be comparing file sizes, QC metrics and the final mutation calls...

No objections for migrating hardcoded paths, especially in separate branch ..

**Comment by HenrikBengtsson** [comment]
@ivan108, there are some `/home/jocostello/LG3/` in the *.pbs files:
```
$ grep -F /home/jocostello/LG3/ *.pbs
Annotate_RNA.pbs:cd /home/jocostello/LG3/exomes/Mutations/ || { echo "ERROR: Failed to set working directory to /home/jocostello/LG3/exomes/Mutations/"; exit 1; }
Annotate_RNA.pbs:	"/home/jocostello/LG3/exomes/Mutations/Patient${PATIENT}.NOR-${NORMAL}__${TYPE}-${TUMOR}.annotated.mutations" \
Annotate_RNA.pbs:	"/home/jocostello/LG3/rna/${RNA}.accepted_hits.bam" \
Germline_only.pbs:cd /home/jocostello/LG3/exomes/Mutations/ || { echo "ERROR: Failed to set working directory to /home/jocostello/LG3/exomes/Mutations/"; exit 1; }
```
and we have:
```sh
$ls -l /home/jocostello/LG3
lrwxrwxrwx 1 jocostello costellolab 41 Aug 27 10:53 /home/jocostello/LG3 -> /costellolab/data1/jocostello/LG3/exomes2
```


Should those `/home/jocostello/LG3/` be replaced by `/costellolab/data1/jocostello/LG3/exomes2/` or `/costellolab/data1/jocostello/LG3/`?

**Comment by HenrikBengtsson** [comment]
Another one:
```sh
$ grep -F /data/jocostello/LG3/ *.pbs
Germline_only.pbs:      "/data/jocostello/LG3/exomes/${PATIENT}/${NORMAL}.bwa.realigned.rmDups.recal.bam" \
Recal_pass2.pbs:mkdir -p "/data/jocostello/LG3/exomes/${PATIENT}"
```
Should `/data/jocostello/LG3/` be updated to `/costellolab/data1/jocostello/LG3/`?

**Comment by HenrikBengtsson** [comment]
FYI, I've prepared code where 98% of the hardcoded paths are replaced by environment variables (with defaults being backward compatible).  This done by scripts that patches the *.pbs and scripts/*.sh files (commits 1653dab & 52f4413).

I'm ready to start testing this, but I hold back until you've completed the tests of the develop branch in Issue #6.

**Comment by ivan108** [comment]
It seems Germline_only.pbs is an older version of Germline.pbs. Both call scripts/Germline.sh,
but older Germline_only.pbs has outdated paths. I suggest to delete Germline_only.pbs to avoid further confusion...just deleted on develop

**Comment by HenrikBengtsson** [comment]
Great. This type of cleanup is one of the things I was hoping for when starting out (and if we delete too much, it's always easy to rollback from the git history).

However, I think we shouldn't do another edits/cleanups to the scripts in the develop branch after the test has been validated.  When the test round is completed, then we should merge what we have into the master branch (=production) so that we set a new milestone.  Then we can start working on the next level of updates.

**Comment by HenrikBengtsson** [comment]
As soon as you consider the test of develop complete, I can do the merge into master.

**Comment by ivan108** [comment]
RE: Annotate_RNA.pbs
Strictly speaking this script is not part of the Exome pipline, I have never used it myself, so I suggest to remove it together with corresponding scripts/Annotate_RNA.sh ?

**Comment by HenrikBengtsson** [comment]
Ok, I've remove them from the 'features/envvars' branch (committed and pushed):
```sh
$ git rm Annotate_RNA.pbs
$ git rm scripts/Annotate_RNA.sh 
```

**Comment by ivan108** [comment]
I did the same on develop..

**Comment by HenrikBengtsson** [comment]
This leaves us only with the following outdated path in one PBS script:
```sh
$ grep -F /data/jocostello/LG3/ *.pbs
Recal_pass2.pbs:mkdir -p "/data/jocostello/LG3/exomes/${PATIENT}"
```
Should `/data/jocostello/LG3/` be updated to `/costellolab/data1/jocostello/LG3/`?

**Comment by HenrikBengtsson** [comment]
> I did the same on develop..
As I wrote before, I don't think we should touch `develop` while running validation tests.  Only after completing the tests.

**Comment by ivan108** [comment]
In general totally I agree, but fixing scripts which we clearly not going to use seems a little counterproductive... :)

**Comment by ivan108** [comment]
RE: Recal_pass2.pbs

Yes, path needs to be updated: from /data/jocostello/LG3/exomes to /costellolab/data1/jocostello/LG3/exomes_recal

**Comment by HenrikBengtsson** [comment]
Thanks, fixed that one occurrence of /data/jocostello/LG3/exomes in feature/envvars (I now see that it was an obvious fix).

**Comment by HenrikBengtsson** [comment]
I've also found
```sh
grep -F patient_ID_conversions.txt scripts/*.sh
scripts/chk_mutdet.sh:conv=/costellolab/data1/mazort/LG3/exome/patient_ID_conversions.txt
```
Is that at all used?  Can I update so it points a file in the current directory (which seems to be what is used):
```sh
conv=patient_ID_conversions.txt
```

**Comment by ivan108** [comment]
Yes, we should use the symbolic link /home/jocostello/shared/LG3_Pipeline/patient_ID_conversions.txt
pointing to actual samples table.

**Comment by HenrikBengtsson** [comment]
STATUS UPDATE 2018-09-09:
I'm chugging along and updating the processing steps such that they run from anywhere and writes to anywhere.  I'm using the barebone test account cbctest2 to validate this.  This far I verified that I can get it to work with:

* [x] `./_run_Trim_P157`
* [x] `./_run_Align_gz_P157`
* [ ] `./_run_Recal_P157_3`

I started the latter 8 hours ago, so it'll take some days before we know.  To get the above working, without rewriting the scripts too much, I've introduced:

* `LG3_HOME`: The absolute path to where the pipeline lives
* `LG3_INPUT_ROOT`: The relative or absolute path to where the input data is (adjusted for each step)
* `LG3_OUTPUT_ROOT`: The relative or absolute path to where the output data should go (typically `output`)

I've also added more file and direction assertions such that we get informative errors as early as possible.

@ivan108, I've updated the scripts such that they are fully backward compatible with how you currently run the production pipeline, i.e. you shouldn't have to change anything - the default will still be as in the past.  I'll soon try to send you some instructions on how to try this new code on your end to verify that this is actually the case.  Stay tuned...

**Comment by HenrikBengtsson** [comment]
I've updated all scripts (PBS, shell, R, and Python) to make use of env vars `LG3_HOME`, `LG3_INPUT_ROOT`, and `LG3_OUTPUT_ROOT` such that it is possible to run the LG3 Pipeline from any account on the TIPCC cluster.

I did this by starting out to update _all_ `*.pbs` and `scripts/*.sh` scripts automatically using various `sed`, `awk`, ... combinations.  In order words, I did modify also those scripts that are outside of Steps 1-7.  I (obviously) did not test those, but the edits "should be" backward compatible.  If I recall correctly, we might clean out those scripts anyway in a later release.

Closing ...
